recipe/patches/0004-patch-zoneinfo_dir_override-to-point-to-our-tzdata.patch

```diff
@@ -0,0 +1,103 @@
+ 1 file changed, 59 insertions(+), 8 deletions(-)
+
+diff --git a/libstdc++-v3/src/c++20/tzdb.cc b/libstdc++-v3/src/c++20/tzdb.cc
+index c7c7cc9de..d8ebffe26 100644
+--- a/libstdc++-v3/src/c++20/tzdb.cc
++++ b/libstdc++-v3/src/c++20/tzdb.cc
+@@ -37,8 +37,12 @@
+ #include <mutex>      // mutex
+ #include <filesystem> // filesystem::read_symlink
+ 
+-#ifndef _AIX
+-# include <cstdlib>   // getenv
++#if defined(__linux__)
++# include <dlfcn.h>   // dladdr
++# include <cstdlib>   // free, malloc, memcpy, realpath
++# include <cstring>   // strlen
++#else defined(_WIN32)
++# include <windows.h>
+ #endif
+ 
+ #if defined __GTHREADS && ATOMIC_POINTER_LOCK_FREE == 2
+@@ -68,13 +72,61 @@ namespace __gnu_cxx
+   // Cannot override weak symbols on AIX.
+   const char* (*zoneinfo_dir_override)() = nullptr;
+ #else
+-  [[gnu::weak]] const char* zoneinfo_dir_override();
+-
+-#if defined(__APPLE__) || defined(__hpux__) \
+-  || (defined(__VXWORKS__) && !defined(__RTP__))
+-  // Need a weak definition for Mach-O et al.
+   [[gnu::weak]] const char* zoneinfo_dir_override()
+   {
++    // get path to library we're in, to determine our location relative to $PREFIX;
++    // with help from the MIT-licensed https://github.com/gpakosz/whereami
++    void* addr = __builtin_extract_return_addr(__builtin_return_address(0));
++    char* this_lib;
++    int i;
++#ifdef _WIN32
++    // we're in %PREFIX%\Library\bin\libstdc++-6.dll
++#   define TO_PREFIX "/../.."
++    // needs single quotes for character (not string) literal
++#   define PATH_SEP '\\'
++    wchar_t buffer[MAX_PATH];
++    HMODULE hm = NULL;
++    // non-zero return means success
++    if (GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS |
++        GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT,
++        (LPCSTR) addr, &hm)) {
++      // returns length of string (not counting null byte), see
++      // https://learn.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulefilenamew#return-value
++      DWORD total_length = GetModuleFileNameW(hm, buffer, sizeof(buffer));
++      if (total_length) {
++        this_lib = (char*)malloc(total_length + 1);
++        memcpy(this_lib, buffer, total_length);
++#else
++    // we're in $PREFIX/lib/libstdcxx.so
++#   define TO_PREFIX "/.."
++#   define PATH_SEP '/'
++    char buffer[PATH_MAX];
++    Dl_info info;
++
++    if (dladdr(addr, &info)) {
++      char* resolved = realpath(info.dli_fname, buffer);
++      if (resolved) {
++        int total_length = (int)strlen(resolved);
++        this_lib = (char*)malloc(total_length + 1);
++        memcpy(this_lib, resolved, total_length);
++#endif
++
++        for (i = (int)total_length - 1; i >= 0; --i) {
++          if (this_lib[i] == PATH_SEP) {
++            // set to null byte so the string ends before the basename
++            this_lib[i] = '\0';
++            break;
++          }
++        }
++        // <string> is included through <chrono>
++        static std::string tz_dir(this_lib);
```

Comment:
I think we better initialize it at the top of this function and return it directly if it was set (i.e. when this functions is subsequently called). What do you think?

Reply:
Yeah, sounds reasonable.

Right now I'm mostly wondering though why the linker cannot find `__gnu_cxx::zoneinfo_file`. It _does_ find `__gnu_cxx::zoneinfo_dir_override` (otherwise the CI for https://github.com/conda-forge/ctng-compilers-feedstock/pull/142/commits/7807b911f42533c59e6dcb69395b8f93a6115b36 wouldn't have passed), but I don't know why it cannot find its immediate neighbour. Possibly because of picking the wrong overload for the various `string_view` specializations between symbol creation and when it's called? 🤷